File: agenkit/safety/permissions.py
"""
Permission-based access control and sandboxing.

Provides:
- Role-Based Access Control (RBAC)
- Permission checks before agent actions
- Sandboxing (allowed paths, commands, operations)
- Resource constraints
"""

from enum import Enum
from typing import Set, Optional, List, Dict, Any
from dataclasses import dataclass, field
from pathlib import Path

from agenkit import Agent, Message


class Permission(Enum):
    """System permissions for agents."""

    # File system
    READ_FILES = "read:files"
    WRITE_FILES = "write:files"
    DELETE_FILES = "delete:files"

    # Command execution
    EXECUTE_COMMANDS = "execute:commands"
    EXECUTE_SHELL = "execute:shell"

    # Database
    QUERY_DATABASE = "query:database"
    WRITE_DATABASE = "write:database"

    # Network
    MAKE_HTTP_REQUESTS = "network:http"
    MAKE_EXTERNAL_API_CALLS = "network:api"

    # System
    MANAGE_USERS = "manage:users"
    MANAGE_AGENTS = "manage:agents"
    ACCESS_SECRETS = "access:secrets"

    # Tools
    USE_TOOLS = "use:tools"
    USE_DANGEROUS_TOOLS = "use:dangerous_tools"


class Role(Enum):
    """Predefined roles with permission sets."""

    # Admin: Full access
    ADMIN = "admin"

    # User: Standard access
    USER = "user"

    # Read-only: View access only
    READONLY = "readonly"

    # Restricted: Minimal access
    RESTRICTED = "restricted"


# Role -> Permissions mapping
ROLE_PERMISSIONS: Dict[Role, Set[Permission]] = {
    Role.ADMIN: {
        Permission.READ_FILES,
        Permission.WRITE_FILES,
        Permission.DELETE_FILES,
        Permission.EXECUTE_COMMANDS,
        Permission.EXECUTE_SHELL,
        Permission.QUERY_DATABASE,
        Permission.WRITE_DATABASE,
        Permission.MAKE_HTTP_REQUESTS,
        Permission.MAKE_EXTERNAL_API_CALLS,
        Permission.MANAGE_USERS,
        Permission.MANAGE_AGENTS,
        Permission.ACCESS_SECRETS,
        Permission.USE_TOOLS,
        Permission.USE_DANGEROUS_TOOLS,
    },
    Role.USER: {
        Permission.READ_FILES,
        Permission.WRITE_FILES,
        Permission.EXECUTE_COMMANDS,
        Permission.QUERY_DATABASE,
        Permission.MAKE_HTTP_REQUESTS,
        Permission.USE_TOOLS,
    },
    Role.READONLY: {
        Permission.READ_FILES,
        Permission.QUERY_DATABASE,
        Permission.USE_TOOLS,  # Basic tool access needed
    },
    Role.RESTRICTED: {
        Permission.READ_FILES,
        Permission.USE_TOOLS,  # Basic tool access needed
    },
}


class PermissionDeniedError(Exception):
    """Raised when permission check fails."""

    def __init__(
        self, message: str, required_permission: Optional[Permission] = None
    ):
        super().__init__(message)
        self.required_permission = required_permission


@dataclass
class Sandbox:
    """
    Defines sandboxed environment for agent execution.

    Specifies:
    - Allowed file paths
    - Allowed commands
    - Allowed database operations
    - Allowed API endpoints
    - Resource limits
    """

    # File system sandbox
    allowed_paths: Set[str] = field(default_factory=set)
    denied_paths: Set[str] = field(default_factory=lambda: {"/etc", "/sys", "/proc"})

    # Command sandbox
    allowed_commands: Set[str] = field(
        default_factory=lambda: {"ls", "cat", "grep", "git", "python"}
    )
    denied_commands: Set[str] = field(
        default_factory=lambda: {"rm", "sudo", "chmod", "chown"}
    )

    # Database sandbox
    allowed_sql_operations: Set[str] = field(
        default_factory=lambda: {"SELECT", "EXPLAIN"}
    )

    # Network sandbox
    allowed_domains: Set[str] = field(default_factory=set)  # Empty = allow all
    denied_domains: Set[str] = field(
        default_factory=lambda: {"localhost", "127.0.0.1", "0.0.0.0"}
    )

    # Resource limits
    max_file_size: int = 10 * 1024 * 1024  # 10MB
    max_execution_time: int = 30  # seconds
    max_memory_mb: int = 512  # MB

    def is_path_allowed(self, path: str) -> tuple[bool, Optional[str]]:
        """
        Check if path is within sandbox.

        Args:
            path: File path to check

        Returns:
            Tuple of (is_allowed, error_message)
        """
        try:
            resolved = Path(path).resolve()

            # Check denied paths first
            for denied in self.denied_paths:
                denied_resolved = Path(denied).resolve()
                try:
                    resolved.relative_to(denied_resolved)
                    return False, f"Path is in denied directory: {denied}"
                except ValueError:
                    pass

            # If allowed_paths specified, must be under one of them
            if self.allowed_paths:
                for allowed in self.allowed_paths:
                    allowed_resolved = Path(allowed).resolve()
                    try:
                        resolved.relative_to(allowed_resolved)
                        return True, None
                    except ValueError:
                        pass

                return False, "Path is outside allowed directories"

            # No allowed_paths specified, just check denied
            return True, None

        except Exception as e:
            return False, f"Path validation error: {e}"

    def is_command_allowed(self, command: str) -> tuple[bool, Optional[str]]:
        """
        Check if command is allowed in sandbox.

        Args:
            command: Command to check (just the command name, not args)

        Returns:
            Tuple of (is_allowed, error_message)
        """
        cmd_name = command.split()[0] if command else ""

        # Check denied commands first
        if cmd_name in self.denied_commands:
            return False, f"Command is denied: {cmd_name}"

        # Check allowed commands
        if self.allowed_commands:
            if cmd_name not in self.allowed_commands:
                return False, f"Command not in allowed list: {cmd_name}"

        return True, None

    def is_sql_operation_allowed(self, sql: str) -> tuple[bool, Optional[str]]:
        """
        Check if SQL operation is allowed.

        Args:
            sql: SQL statement

        Returns:
            Tuple of (is_allowed, error_message)
        """
        operation = sql.strip().upper().split()[0] if sql else ""

        if operation not in self.allowed_sql_operations:
            return False, f"SQL operation not allowed: {operation}"

        return True, None

    def is_domain_allowed(self, domain: str) -> tuple[bool, Optional[str]]:
        """
        Check if domain is allowed for network requests.

        Args:
            domain: Domain to check

        Returns:
            Tuple of (is_allowed, error_message)
        """
        # Check denied domains first
        if domain in self.denied_domains:
            return False, f"Domain is denied: {domain}"

        # If allowed_domains specified, must be in list
        if self.allowed_domains:
            if domain not in self.allowed_domains:
                return False, f"Domain not in allowed list: {domain}"

        return True, None


class PermissionMiddleware(Agent):
    """
    Middleware for permission checks and sandboxing.

    Enforces:
    - Role-based permissions
    - Sandbox constraints
    - Resource limits

    Usage:
        sandbox = Sandbox(
            allowed_paths={"/app/data"},
            allowed_commands={"git", "ls", "cat"}
        )

        agent = PermissionMiddleware(
            base_agent,
            role=Role.USER,
            sandbox=sandbox
        )
    """

    def __init__(
        self,
        agent: Agent,
        role: Role = Role.USER,
        custom_permissions: Optional[Set[Permission]] = None,
        sandbox: Optional[Sandbox] = None,
    ):
        """
        Initialize permission middleware.

        Args:
            agent: Agent to wrap
            role: User role (determines permissions)
            custom_permissions: Custom permission set (overrides role)
            sandbox: Sandbox constraints
        """
        self._agent = agent
        self.role = role
        self.permissions = custom_permissions or ROLE_PERMISSIONS.get(
            role, set()
        )
        self.sandbox = sandbox or Sandbox()

    @property
    def name(self) -> str:
        """Return the name of the underlying agent."""
        return self._agent.name

    @property
    def capabilities(self) -> list[str]:
        """Return capabilities of the underlying agent."""
        return self._agent.capabilities

    def has_permission(self, permission: Permission) -> bool:
        """Check if agent has permission."""
        return permission in self.permissions

    def check_permission(self, permission: Permission) -> None:
        """
        Check permission and raise error if denied.

        Args:
            permission: Required permission

        Raises:
            PermissionDeniedError: If permission not granted
        """
        if not self.has_permission(permission):
            raise PermissionDeniedError(
                f"Permission denied: {permission.value} required (role: {self.role.value})",
                required_permission=permission,
            )

    async def process(self, message: Message) -> Message:
        """
        Process with permission checks.

        Note: This middleware checks for general USE_TOOLS permission.
        Specific permission checks should be done by tools themselves
        or by extending this middleware.
        """
        # Basic permission check
        self.check_permission(Permission.USE_TOOLS)

        # Check for dangerous operations in message content
        content_str = str(message.content).lower() if message.content else ""

        # Detect file operations
        if any(keyword in content_str for keyword in ["read file", "write file", "delete file"]):
            if "delete" in content_str:
                self.check_permission(Permission.DELETE_FILES)
            elif "write" in content_str:
                self.check_permission(Permission.WRITE_FILES)
            else:
                self.check_permission(Permission.READ_FILES)

        # Detect command execution
        if any(keyword in content_str for keyword in ["execute", "run command", "shell"]):
            if "shell" in content_str:
                self.check_permission(Permission.EXECUTE_SHELL)
            else:
                self.check_permission(Permission.EXECUTE_COMMANDS)

        # Detect database operations
        # Check for write operations first (more specific)
        if any(op in content_str for op in ["insert", "update", "delete", "drop", "alter", "create table"]):
            # Likely a SQL write operation
            self.check_permission(Permission.WRITE_DATABASE)
        elif any(keyword in content_str for keyword in ["query", "database", "sql", "select", "from"]):
            # Database read operation
            self.check_permission(Permission.QUERY_DATABASE)

        # Process with wrapped agent
        return await self._agent.process(message)


def permissions(
    role: Role = Role.USER,
    custom_permissions: Optional[Set[Permission]] = None,
    sandbox: Optional[Sandbox] = None,
):
    """
    Create permission middleware function.

    Args:
        role: User role
        custom_permissions: Custom permissions (overrides role)
        sandbox: Sandbox constraints

    Returns:
        Middleware function

    Usage:
        sandbox = Sandbox(allowed_paths={"/app/data"})

        agent = applyMiddleware(base_agent, [
            permissions(role=Role.USER, sandbox=sandbox),
            retry(),
        ])
    """
    def middleware(agent: Agent) -> Agent:
        return PermissionMiddleware(agent, role, custom_permissions, sandbox)

    return middleware

File: agenkit-ts/src/safety/permissions.ts
/**
 * Permission-based access control and sandboxing.
 *
 * Provides:
 * - Role-Based Access Control (RBAC)
 * - Permission checks before agent actions
 * - Sandboxing (allowed paths, commands, operations)
 * - Resource constraints
 */

import * as path from 'path';
import { Agent, Message } from '../core/interfaces';

/**
 * System permissions for agents.
 */
export enum Permission {
  // File system
  READ_FILES = 'read:files',
  WRITE_FILES = 'write:files',
  DELETE_FILES = 'delete:files',

  // Command execution
  EXECUTE_COMMANDS = 'execute:commands',
  EXECUTE_SHELL = 'execute:shell',

  // Database
  QUERY_DATABASE = 'query:database',
  WRITE_DATABASE = 'write:database',

  // Network
  MAKE_HTTP_REQUESTS = 'network:http',
  MAKE_EXTERNAL_API_CALLS = 'network:api',

  // System
  MANAGE_USERS = 'manage:users',
  MANAGE_AGENTS = 'manage:agents',
  ACCESS_SECRETS = 'access:secrets',

  // Tools
  USE_TOOLS = 'use:tools',
  USE_DANGEROUS_TOOLS = 'use:dangerous_tools',
}

/**
 * Predefined roles with permission sets.
 */
export enum Role {
  ADMIN = 'admin',
  USER = 'user',
  READONLY = 'readonly',
  RESTRICTED = 'restricted',
}

/**
 * Role to permissions mapping.
 */
export const ROLE_PERMISSIONS: Record<Role, Set<Permission>> = {
  [Role.ADMIN]: new Set([
    Permission.READ_FILES,
    Permission.WRITE_FILES,
    Permission.DELETE_FILES,
    Permission.EXECUTE_COMMANDS,
    Permission.EXECUTE_SHELL,
    Permission.QUERY_DATABASE,
    Permission.WRITE_DATABASE,
    Permission.MAKE_HTTP_REQUESTS,
    Permission.MAKE_EXTERNAL_API_CALLS,
    Permission.MANAGE_USERS,
    Permission.MANAGE_AGENTS,
    Permission.ACCESS_SECRETS,
    Permission.USE_TOOLS,
    Permission.USE_DANGEROUS_TOOLS,
  ]),
  [Role.USER]: new Set([
    Permission.READ_FILES,
    Permission.WRITE_FILES,
    Permission.EXECUTE_COMMANDS,
    Permission.QUERY_DATABASE,
    Permission.MAKE_HTTP_REQUESTS,
    Permission.USE_TOOLS,
  ]),
  [Role.READONLY]: new Set([
    Permission.READ_FILES,
    Permission.QUERY_DATABASE,
    Permission.USE_TOOLS,
  ]),
  [Role.RESTRICTED]: new Set([Permission.READ_FILES, Permission.USE_TOOLS]),
};

/**
 * Error thrown when permission check fails.
 */
export class PermissionDeniedError extends Error {
  constructor(message: string, public readonly requiredPermission?: Permission) {
    super(message);
    this.name = 'PermissionDeniedError';
  }
}

/**
 * Defines sandboxed environment for agent execution.
 *
 * Specifies:
 * - Allowed file paths
 * - Allowed commands
 * - Allowed database operations
 * - Allowed API endpoints
 * - Resource limits
 *
 * Example:
 *   const sandbox = new Sandbox({
 *     allowedPaths: new Set(['/app/data']),
 *     allowedCommands: new Set(['git', 'ls', 'cat']),
 *     maxFileSize: 10 * 1024 * 1024, // 10MB
 *   });
 *
 *   const [isAllowed, errorMsg] = sandbox.isPathAllowed('/app/data/file.txt');
 */
export class Sandbox {
  // File system sandbox
  private allowedPaths: Set<string>;
  private deniedPaths: Set<string>;

  // Command sandbox
  private allowedCommands: Set<string>;
  private deniedCommands: Set<string>;

  // Database sandbox
  private allowedSqlOperations: Set<string>;

  // Network sandbox
  private allowedDomains: Set<string>;
  private deniedDomains: Set<string>;

  // Resource limits
  readonly maxFileSize: number;
  readonly maxExecutionTime: number;
  readonly maxMemoryMb: number;

  constructor(config?: {
    allowedPaths?: Set<string>;
    deniedPaths?: Set<string>;
    allowedCommands?: Set<string>;
    deniedCommands?: Set<string>;
    allowedSqlOperations?: Set<string>;
    allowedDomains?: Set<string>;
    deniedDomains?: Set<string>;
    maxFileSize?: number;
    maxExecutionTime?: number;
    maxMemoryMb?: number;
  }) {
    this.allowedPaths = config?.allowedPaths ?? new Set();
    this.deniedPaths = config?.deniedPaths ?? new Set(['/etc', '/sys', '/proc']);

    this.allowedCommands = config?.allowedCommands ?? new Set(['ls', 'cat', 'grep', 'git', 'python']);
    this.deniedCommands = config?.deniedCommands ?? new Set(['rm', 'sudo', 'chmod', 'chown']);

    this.allowedSqlOperations = config?.allowedSqlOperations ?? new Set(['SELECT', 'EXPLAIN']);

    this.allowedDomains = config?.allowedDomains ?? new Set(); // Empty = allow all
    this.deniedDomains = config?.deniedDomains ?? new Set(['localhost', '127.0.0.1', '0.0.0.0']);

    this.maxFileSize = config?.maxFileSize ?? 10 * 1024 * 1024; // 10MB
    this.maxExecutionTime = config?.maxExecutionTime ?? 30; // seconds
    this.maxMemoryMb = config?.maxMemoryMb ?? 512; // MB
  }

  /**
   * Check if path is within sandbox.
   *
   * Returns tuple of (is_allowed, error_message)
   */
  isPathAllowed(filePath: string): [boolean, string | null] {
    try {
      const resolved = path.resolve(filePath);

      // Check denied paths first
      for (const denied of this.deniedPaths) {
        const deniedResolved = path.resolve(denied);
        if (resolved.startsWith(deniedResolved)) {
          return [false, `Path is in denied directory: ${denied}`];
        }
      }

      // If allowed_paths specified, must be under one of them
      if (this.allowedPaths.size > 0) {
        for (const allowed of this.allowedPaths) {
          const allowedResolved = path.resolve(allowed);
          if (resolved.startsWith(allowedResolved)) {
            return [true, null];
          }
        }

        return [false, 'Path is outside allowed directories'];
      }

      // No allowed_paths specified, just check denied
      return [true, null];
    } catch (error) {
      return [false, `Path validation error: ${error}`];
    }
  }

  /**
   * Check if command is allowed in sandbox.
   *
   * Returns tuple of (is_allowed, error_message)
   */
  isCommandAllowed(command: string): [boolean, string | null] {
    const cmdName = command.trim().split(/\s+/)[0] || '';

    // Check denied commands first
    if (this.deniedCommands.has(cmdName)) {
      return [false, `Command is denied: ${cmdName}`];
    }

    // Check allowed commands
    if (this.allowedCommands.size > 0 && !this.allowedCommands.has(cmdName)) {
      return [false, `Command not in allowed list: ${cmdName}`];
    }

    return [true, null];
  }

  /**
   * Check if SQL operation is allowed.
   *
   * Returns tuple of (is_allowed, error_message)
   */
  isSqlOperationAllowed(sql: string): [boolean, string | null] {
    const operation = sql.trim().toUpperCase().split(/\s+/)[0] || '';

    if (!this.allowedSqlOperations.has(operation)) {
      return [false, `SQL operation not allowed: ${operation}`];
    }

    return [true, null];
  }

  /**
   * Check if domain is allowed for network requests.
   *
   * Returns tuple of (is_allowed, error_message)
   */
  isDomainAllowed(domain: string): [boolean, string | null] {
    // Check denied domains first
    if (this.deniedDomains.has(domain)) {
      return [false, `Domain is denied: ${domain}`];
    }

    // If allowed_domains specified, must be in list
    if (this.allowedDomains.size > 0 && !this.allowedDomains.has(domain)) {
      return [false, `Domain not in allowed list: ${domain}`];
    }

    return [true, null];
  }
}

/**
 * Middleware for permission checks and sandboxing.
 *
 * Enforces:
 * - Role-based permissions
 * - Sandbox constraints
 * - Resource limits
 *
 * Example:
 *   const sandbox = new Sandbox({
 *     allowedPaths: new Set(['/app/data']),
 *     allowedCommands: new Set(['git', 'ls', 'cat']),
 *   });
 *
 *   const agent = new PermissionMiddleware(
 *     baseAgent,
 *     Role.USER,
 *     undefined, // no custom permissions
 *     sandbox,
 *   );
 */
export class PermissionMiddleware implements Agent {
  readonly name: string;
  readonly capabilities?: string[];

  private agent: Agent;
  private role: Role;
  private permissions: Set<Permission>;
  private sandbox: Sandbox;

  constructor(
    agent: Agent,
    role: Role = Role.USER,
    customPermissions?: Set<Permission>,
    sandbox?: Sandbox,
  ) {
    this.agent = agent;
    this.name = agent.name;
    this.capabilities = agent.capabilities;
    this.role = role;
    this.permissions = customPermissions || ROLE_PERMISSIONS[role] || new Set();
    this.sandbox = sandbox || new Sandbox();
  }

  /**
   * Check if agent has permission.
   */
  hasPermission(permission: Permission): boolean {
    return this.permissions.has(permission);
  }

  /**
   * Check permission and raise error if denied.
   */
  checkPermission(permission: Permission): void {
    if (!this.hasPermission(permission)) {
      throw new PermissionDeniedError(
        `Permission denied: ${permission} required (role: ${this.role})`,
        permission,
      );
    }
  }

  /**
   * Process message with permission checks.
   *
   * Note: This middleware checks for general USE_TOOLS permission.
   * Specific permission checks should be done by tools themselves
   * or by extending this middleware.
   */
  async process(message: Message): Promise<Message> {
    // Basic permission check
    this.checkPermission(Permission.USE_TOOLS);

    // Check for dangerous operations in message content
    const contentStr = message.content ? String(message.content).toLowerCase() : '';

    // Detect file operations
    if (
      contentStr.includes('read file') ||
      contentStr.includes('write file') ||
      contentStr.includes('delete file')
    ) {
      if (contentStr.includes('delete')) {
        this.checkPermission(Permission.DELETE_FILES);
      } else if (contentStr.includes('write')) {
        this.checkPermission(Permission.WRITE_FILES);
      } else {
        this.checkPermission(Permission.READ_FILES);
      }
    }

    // Detect command execution
    if (
      contentStr.includes('execute') ||
      contentStr.includes('run command') ||
      contentStr.includes('shell')
    ) {
      if (contentStr.includes('shell')) {
        this.checkPermission(Permission.EXECUTE_SHELL);
      } else {
        this.checkPermission(Permission.EXECUTE_COMMANDS);
      }
    }

    // Detect database operations
    // Check for write operations first (more specific)
    const dbWriteKeywords = ['insert', 'update', 'delete', 'drop', 'alter', 'create table'];
    if (dbWriteKeywords.some((kw) => contentStr.includes(kw))) {
      // Likely a SQL write operation
      this.checkPermission(Permission.WRITE_DATABASE);
    } else if (
      contentStr.includes('query') ||
      contentStr.includes('database') ||
      contentStr.includes('sql') ||
      contentStr.includes('select') ||
      contentStr.includes('from')
    ) {
      // Database read operation
      this.checkPermission(Permission.QUERY_DATABASE);
    }

    // Process with wrapped agent
    return await this.agent.process(message);
  }
}

/**
 * Create permission middleware function.
 *
 * Example:
 *   const sandbox = new Sandbox({ allowedPaths: new Set(['/app/data']) });
 *
 *   const agent = applyMiddleware(baseAgent, [
 *     permissions({
 *       role: Role.USER,
 *       sandbox,
 *     }),
 *   ]);
 */
export function permissions(config?: {
  role?: Role;
  customPermissions?: Set<Permission>;
  sandbox?: Sandbox;
}): (agent: Agent) => Agent {
  return (agent: Agent) =>
    new PermissionMiddleware(agent, config?.role, config?.customPermissions, config?.sandbox);
}
